src/details/ArborX_DetailsHalfTraversal.hpp

```diff
@@ -53,30 +53,59 @@ struct HalfTraversal
 
   KOKKOS_FUNCTION void operator()(int i) const
   {
-    auto const predicate =
-        _get_predicate(HappyTreeFriends::getBoundingVolume(_bvh, i));
+    auto const leaf_i = HappyTreeFriends::leafIndex(_bvh, i);
+    auto const predicate = _get_predicate(
+        HappyTreeFriends::getBoundingVolume(LeafNodeTag{}, _bvh, leaf_i));
     auto const leaf_permutation_i =
-        HappyTreeFriends::getLeafPermutationIndex(_bvh, i);
+        HappyTreeFriends::getLeafPermutationIndex(_bvh, leaf_i);
 
-    int node = HappyTreeFriends::getRope(_bvh, i);
+    int node = HappyTreeFriends::getRope(LeafNodeTag{}, _bvh, leaf_i);
     while (node != ROPE_SENTINEL)
     {
-      if (predicate(HappyTreeFriends::getBoundingVolume(_bvh, node)))
+      auto const leaf_node = HappyTreeFriends::leafIndex(_bvh, node);
+      bool const is_leaf = HappyTreeFriends::isLeaf(_bvh, node);
+
+      bool satisfied = false;
+      constexpr bool same_bounding_volume =
+          std::is_same_v<decltype(HappyTreeFriends::getBoundingVolume(
+                             LeafNodeTag{}, std::declval<BVH>(), 0)),
+                         decltype(HappyTreeFriends::getBoundingVolume(
+                             InternalNodeTag{}, std::declval<BVH>(), 0))>;
+      if constexpr (same_bounding_volume)
+      {
+        satisfied = predicate((is_leaf ? HappyTreeFriends::getBoundingVolume(
+                                             LeafNodeTag{}, _bvh, leaf_node)
+                                       : HappyTreeFriends::getBoundingVolume(
+                                             InternalNodeTag{}, _bvh, node)));
+      }
+      else
+      {
+        if (is_leaf)
+          satisfied = predicate(HappyTreeFriends::getBoundingVolume(
+              LeafNodeTag{}, _bvh, leaf_node));
+        else
+          satisfied = predicate(HappyTreeFriends::getBoundingVolume(
+              InternalNodeTag{}, _bvh, node));
```

Comment:
These `else` conditions for different bounding volumes are more for the future, and not used right now. I would be fine replacing them with a `static_assert` on `same_bounding_volume`. But they will have to be added back anyway soon, so I thought I'd do it here.